### ISSUE TITLE
8bpc PNGs not working

### ISSUE THREAD

**Issue by rickomax** [StbSharp/StbImageSharp#3]
Hi!

I have using this project for a while and its a great piece of work!
I had some issues today while loading paletted (8bpc) PNGs.

Have you ever tested them?

I have attached the original image (1) and the result I get with StbImageSharp (2).

![1](https://user-images.githubusercontent.com/12863984/91670092-9751ea00-eaf0-11ea-8082-ac47e8effc40.png)
![2](https://user-images.githubusercontent.com/12863984/91670107-b94b6c80-eaf0-11ea-83b3-e1eb0e79f5a7.png)

**Comment by rds1983** [comment]
Works for me. I've made a small MG project(attached) that loads the image using StbImageSharp and draws. And it works as expected.
![image](https://user-images.githubusercontent.com/1057289/91675439-96f43600-eb66-11ea-87a8-f8707c6e2072.png)


Though maybe github spoiled that png. Or maybe Chrome did when I saved it.

[StbTest.zip](https://github.com/StbSharp/StbImageSharp/files/5147929/StbTest.zip)

**Comment by rickomax** [comment]
Hey!
Thank you for the demo.
It should be something into the code I've changed then.
I had to change the code a bit to write to Unity pre-allocated NativeArrays instead of allocating memory with marshal.
I had to change the stbi__convert_format code which should be causing the issue.
I will investigate.
I plan to release a branch with the code adapted to write straight into Unity Texture2Ds without extra memory allocation :)

**Comment by rds1983** [comment]
Nice,
Then you may be interested in taking a look at this fork: https://github.com/TechnologicalPizza/StbSharp.Image
TechnologicalPizza made incredible amount of work and turned original unreadable auto-translated code into the quite readable one. Moreover I think he made it so the image is read into Span<byte>. Unsure though if Unity3D supports Span yet.

**Comment by rickomax** [comment]
Wow!
I will check it out.
Well, there is a single way to load textures in Unity without letting it do a memory copy of all the pixel buffer.
The texture should be created on the main thread, then the command GetRawPixelData will return a NativeArray formatted with a user-selected data-type. I load my textures on a parallel thread using StbImageSharp, then every time it has to generate the pixel-data buffer, I dispatch a call to the main thread, letting Unity create the texture with the right size, then when the texture has been created, I continue the parallel thread execution normally and let StbImageSharp fill the data buffer retrieved from the texture created on the main thread.
If I assign a byte[] to a texture or even a NativeArray to a texture in Unity, it will do a memory copy as it always allocates memory when creating a texture, so I had to create the approach above.

This is the project where I'm using StbImageSharp (proper credits are given on the package):
https://twitter.com/realrickomax/status/1300277358813904899

**Comment by rds1983** [comment]
Cool!
I wonder, why stock Unity3D image loading(Texture.LoadImage) wasnt good?

**Comment by rickomax** [comment]
It accepts PNG and JPEG files only and if I want it to generate mip-maps automatically, I had to do some tricks with RenderTextures which requires duplicating the texture raw data, something I want to avoid. So since I can control how the Texture2D is created using my approach, I create it passing "true" to the mip-maps parameter and I don't touch the mip-map reserved byte-area and let Unity calculate the mip-maps automatically when calling Texture2D.Apply.

**Comment by rds1983** [comment]
I see, thanks for the explanation.

**Comment by rickomax** [comment]
Hi!

I've checked out my code and discovered that it was an issue in my end.
I've changed methods like `stbi__expand_png_palette` that does extra memory allocation to copy the output data to the Unity Texture NativeArray, but I forgot to do that on the palette method above, so it was using the "not-expanded" data on the final result.

I'll publish my branch when I have all these issues fixed.

Thanks for the excellent library!
